source/common/stats/deferred_creation.h

```diff
@@ -0,0 +1,148 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/stats/scope.h"
+#include "envoy/stats/stats.h"
+
+#include "source/common/common/cleanup.h"
+#include "source/common/common/thread.h"
+#include "source/common/stats/symbol_table.h"
+#include "source/common/stats/utility.h"
+
+namespace Envoy {
+namespace Stats {
+
+template <typename StatsStructType> class DirectStats;
+template <typename StatsStructType> class DeferredCreation;
+
+/**
+ * Interface for stats lazy initialization.
+ * To reduce memory and CPU consumption, Envoy can enable the bootstrap config
+ * :ref:`enable_deferred_creation_stats
+ * <envoy_v3_api_field_config.bootstrap.v3.Bootstrap.enable_deferred_creation_stats>`.
+ * A 'StatsStructType' is only created when any of its field is referenced.
+ * See more context: https://github.com/envoyproxy/envoy/issues/23575
+ */
+template <typename StatsStructType> class DeferredCreationCompatibleInterface {
```

Comment:
I think this class definition should be moved to envoy/stats/stats.h, since it is defining an interface.

I think that might enable you to remove a bunch of changes in this PR that look strange to me -- like they shouldn't really be necessary. An example is source/common/http/codec_client.h .

Reply:
I am not sure that's gonna work, we need to merge the whole file into envoy/stats/stat.h if we want to remove the header. 

right now, the codec_client.h is actually looking for the DeferredCreationCompatibleStats definition because of the "->" reference. Compiler will fail if we remove the header.